lib/time_interval.rb

```diff
@@ -0,0 +1,29 @@
+require "date"
+
+module Hotel
+  class TimeInterval
+    attr_reader :check_in, :check_out
+
+    def initialize(check_in, check_out)
+      if check_in >= check_out
+        raise ArgumentError.new("Check in time cannot be greater than or equal to check out time")
+      end
+
+      @check_in = check_in
+      @check_out = check_out
+    end
+
+    def overlap?(time_interval)
+      return !(time_interval.check_out <= @check_in ||
+               @check_out <= time_interval.check_in)
+    end
+
+    def has_date?(date)
+      return date >= @check_in && date < @check_out
+    end
+
+    def equals?(time_interval)
```

Comment:
You could override `==` here instead of defining your own `equals?` method to clean up your code a tiny bit.

```suggestion
    def ==(time_interval)
```